# Preparation

Preparation biasa dilakukan untuk mempersiapkan data sebelum masuk dalam tahap pemodelan. <br>
Berikut adalah tahapan yang akan dilalui pada data `SC_HW1_bank_data.csv` sebelum tahap pemodelan :
1. Import Library
2. Input Dataset
3. Preprocessing
4. Train-Test Split

## Import Library

In [3]:
import pandas as pd
import numpy as np

## Input Dataset

In [4]:
#Membaca data dan memasukkannya ke dalam bentuk Pandas Dataframe
df = pd.read_csv('https://raw.githubusercontent.com/Rietaros/kampus_merdeka/main/SC_HW1_bank_data.csv')

In [5]:
#Jalankan code untuk mengecek nama kolom yang tersedia
df.columns

Index(['RowNumber', 'CustomerId', 'Geography', 'Gender', 'Age', 'Tenure',
       'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember',
       'EstimatedSalary', 'Exited'],
      dtype='object')

In [6]:
#Hilangkan kolom yang dirasa tidak relevan dengan model (contoh: ID). None dapat diisi dengan nama-nama kolom yang akan digunakan.
#Contoh df = df[['X1','X2', 'Y']].copy()

#START CODE
df = df.iloc[:, 2:].copy()

## Preprocessing

In [7]:
#Lakukan One-Hot Encoder untuk data categorical, dengan fungsi pandas get_dummies

#START CODE
df = pd.get_dummies(df)
df.head()

Age  Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0   42       2       0.00              1          1               1   
1   41       1   83807.86              1          0               1   
2   42       8  159660.80              3          1               0   
3   39       1       0.00              2          0               0   
4   43       2  125510.82              1          1               1   

   EstimatedSalary  Exited  Geography_France  Geography_Germany  \
0        101348.88       1                 1                  0   
1        112542.58       0                 0                  0   
2        113931.57       1                 1                  0   
3         93826.63       0                 1                  0   
4         79084.10       0                 0                  0   

   Geography_Spain  Gender_Female  Gender_Male  
0                0              1            0  
1                1              1            0  
2                0              1            0  
3                0              1            0  
4                1              1            0

In [8]:
#Pisahkan mana X (feature) dengan Y,
#Y adalah kolom "Exited"

#START CODE
temp1 = df.iloc[:, :7]
temp2 = df.iloc[:, 8:]
X = pd.concat([temp1, temp2], axis=1)
y = df['Exited']

In [9]:
# lakukan Scaler dan/atau Noermalisasi jika diperlukan
from sklearn.preprocessing import MinMaxScaler, StandardScaler

# START CODE
scaler = StandardScaler()
X_transform = scaler.fit_transform(X)

In [10]:
# ini digunakan jika dilakukan scaler/Normalisas. Jika tidak, code ini bisa dilewat dan diganti dengan code yang ada di dalam komen
X_transform = pd.DataFrame(X_transform, columns = X.columns)
#X_transform = X.copy()

## Train-Test Split

In [11]:
#Split menjadi train dan test dengan test_size 25%
#Tidak perlu mengubah code ini

from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X_transform,y,test_size = 0.25,random_state = 123)

# Modeling

## Model1

### Soal :
Jelaskan secara Singkat Model pertama yang digunakan!

jawab
K-Nearest Neighbors (K-NN) Classifier adalah salah satu algoritma klasifikasi dalam pemodelan statistik dan pembelajaran mesin. Ini adalah metode klasifikasi yang sederhana dan intuitif yang digunakan untuk mengklasifikasikan objek berdasarkan sejumlah tetangga terdekat dalam ruang fitur. K-NN bekerja dengan prinsip bahwa objek yang serupa cenderung berada dalam jarak yang dekat satu sama lain dalam ruang fitur.

In [12]:
# K-nearest Neighbor
# START CODE
from sklearn.neighbors import KNeighborsClassifier
model1 = KNeighborsClassifier()
params1 = {'n_neighbors' : [3, 5, 7, 9, 13]}
# END CODE

# lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
from sklearn.model_selection import GridSearchCV
grid = GridSearchCV(
             estimator= model1,
             param_grid= params1,
             scoring = 'accuracy',
             n_jobs = 10, # core cpu yang digunakan
             cv = 10 # 3-fold cross validation (artinya kita melakukan iterasi model sebanyak 3 kali)
            )

grid.fit(X_train,y_train)
grid.best_params_

{'n_neighbors': 13}

In [13]:
#lakukan evaluasi dengan beberapa matric di bawah ini
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.85      0.97      0.90      1983
           1       0.73      0.34      0.46       517

    accuracy                           0.84      2500
   macro avg       0.79      0.65      0.68      2500
weighted avg       0.82      0.84      0.81      2500


[[1917   66]
 [ 342  175]]

0.8368


## Model2
### Soal :
Jelaskan secara Singkat Model ke-2 yang digunakan!

###jawab
Decision Tree (Pohon Keputusan) adalah salah satu algoritma yang digunakan dalam pemodelan dan pembelajaran mesin. Ini adalah model prediktif yang menggambarkan struktur pohon dengan keputusan dan konsekuensi berbagai situasi. Decision Tree digunakan baik untuk masalah klasifikasi maupun regresi.

In [17]:
# Decision Tree Model
# START CODE
from sklearn.tree import DecisionTreeClassifier
model2 = DecisionTreeClassifier()
params2 = {'criterion': ['gini', 'entropy'],
           'max_depth': [8, 12, 14],
           'random_state': [24, 42]}
# END CODE

# lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
from sklearn.model_selection import GridSearchCV
grid2 = GridSearchCV(
             estimator= model2,
             param_grid= params2,
             scoring = 'accuracy',
             n_jobs = 10, # core cpu yang digunakan
             cv = 10 # 3-fold cross validation (artinya kita melakukan iterasi model sebanyak 3 kali)
            )

grid2.fit(X_train,y_train)
grid2.best_params_

{'criterion': 'entropy', 'max_depth': 8, 'random_state': 24}

In [18]:
# lakukan evaluasi
y_pred2 = grid2.predict(X_test)

print(classification_report(y_test,y_pred2))
print("")
print(confusion_matrix(y_test,y_pred2))
print("")
print(accuracy_score(y_test, y_pred2))

              precision    recall  f1-score   support

           0       0.87      0.96      0.91      1983
           1       0.73      0.46      0.56       517

    accuracy                           0.85      2500
   macro avg       0.80      0.71      0.74      2500
weighted avg       0.84      0.85      0.84      2500


[[1896   87]
 [ 281  236]]

0.8528


## Model3
### Soal :
Jelaskan secara Singkat Model ke-3 yang digunakan!

### Jawab
Random Forest adalah salah satu algoritma pemodelan dalam pembelajaran mesin yang digunakan untuk tugas klasifikasi dan regresi. Ini adalah metode ensemble, yang berarti menggabungkan beberapa model untuk meningkatkan kinerja dan akurasi prediksi.

In [16]:
# Random Forest Model
# START CODE
from sklearn.ensemble import RandomForestClassifier
model3 = RandomForestClassifier()
params3 = {'criterion': ['gini', 'entropy'],
          'max_depth': [8, 12, 14],
          'random_state': [24, 42],
          'n_estimators': [10, 50, 100]}
# END CODE

# lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
from sklearn.model_selection import GridSearchCV
grid3 = GridSearchCV(
             estimator= model3,
             param_grid= params3,
             scoring = 'accuracy',
             n_jobs = 10, # core cpu yang digunakan
             cv = 10 # 3-fold cross validation (artinya kita melakukan iterasi model sebanyak 3 kali)
            )

grid3.fit(X_train,y_train)
grid3.best_params_

{'criterion': 'entropy',
 'max_depth': 12,
 'n_estimators': 100,
 'random_state': 24}

In [19]:
# lakukan evaluasi
y_pred3 = grid3.predict(X_test)

print(classification_report(y_test,y_pred3))
print("")
print(confusion_matrix(y_test,y_pred3))
print("")
print(accuracy_score(y_test, y_pred3))

              precision    recall  f1-score   support

           0       0.87      0.97      0.92      1983
           1       0.78      0.46      0.58       517

    accuracy                           0.86      2500
   macro avg       0.83      0.71      0.75      2500
weighted avg       0.85      0.86      0.85      2500


[[1918   65]
 [ 281  236]]

0.8616


## Tarik Kesimpulan Model Mana yang terbaik beserta alasannya

##Jawab
Hasil eksperimen ketiga model di atas menunjukkan bahwa **Random Forest adalah pilihan terbaik dalam memproyeksikan pelanggan bank yang tidak aktif**. Berdasarkan eksperimen ini, **Random Forest memiliki tingkat akurasi sebesar 86.16%**, yang secara signifikan lebih tinggi daripada dua model lainnya yang telah diuji sebelumnya. Dengan kata lain, Random Forest mampu memberikan prediksi yang lebih akurat dalam mengidentifikasi pelanggan bank yang tidak aktif dibandingkan dengan K-Nearest Neighbors (KNN) dan Decision Tree.